src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+					volumeConfig ?= {}
+					volumeConfig.labels ?= {}
+					return volumeConfig
+				Promise.map(JSON.parse(app.services), (service) => @createTargetService(service, configOpts))
+				.then (services) ->
+					# If a named volume is defined in a service, we add it app-wide so that we can track it and purge it
+					_.forEach services, (s) ->
+						serviceNamedVolumes = s.getNamedVolumes()
+						_.forEach serviceNamedVolumes, (name) ->
+							volumes[name] ?= { labels: {} }
+					outApp = {
+						appId: app.appId
+						name: app.name
+						commit: app.commit
+						releaseId: app.releaseId
+						services: services
+						networks: JSON.parse(app.networks)
+						volumes: volumes
+					}
+					return outApp
+		)
+
+	setTarget: (apps, dependent , trx) =>
+		setInTransaction = (trx) =>
+			Promise.try =>
+				if apps?
+					appsArray = _.map apps, (app, appId) ->
+						appClone = _.clone(app)
+						appClone.appId = appId
+						return appClone
+					Promise.map(appsArray, @normaliseAppForDB)
+					.then (appsForDB) =>
+						Promise.map appsForDB, (app) =>
+							@db.upsertModel('app', app, { appId: app.appId }, trx)
+						.then ->
+							trx('app').whereNotIn('appId', _.map(appsForDB, 'appId')).del()
+			.then =>
+				@proxyvisor.setTargetInTransaction(dependent, trx)
+
+		Promise.try =>
+			if trx?
+				setInTransaction(trx)
+			else
+				@db.transaction(setInTransaction)
+		.then =>
+			@_targetVolatilePerServiceId = {}
+
+	setTargetVolatileForService: (serviceId, target) ->
+		@_targetVolatilePerServiceId[serviceId] ?= {}
+		_.assign(@_targetVolatilePerServiceId, target)
+
+	getTargetApps: =>
+		Promise.map(@db.models('app').select(), @normaliseAndExtendAppFromDB)
+		.map (app) =>
+			if !_.isEmpty(app.services)
+				app.services = _.map app.services, (service) =>
+					_.merge(service, @_targetVolatilePerServiceId[service.serviceId]) if @_targetVolatilePerServiceId[service.serviceId]?
+					return service
+			return app
+
+	getDependentTargets: =>
+		@proxyvisor.getTarget()
+
+	bestDeltaSource: (image, available) ->
+		if !image.dependent
+			for availableImage in available
+				if availableImage.serviceName == image.serviceName and availableImage.appId == image.appId
+					return availableImage.name
+			for availableImage in available
+				if availableImage.serviceName == image.serviceName
+					return availableImage.name
+		for availableImage in available
+			if availableImage.appId == image.appId
+				return availableImage.name
+		return 'resin/scratch'
+
+	# return images that:
+	# - are not used in the current state, and
+	# - are not going to be used in the target state, and
+	# - are not needed for delta source / pull caching or would be used for a service with delete-then-download as strategy
+	_unnecessaryImages: (current, target, available) =>
+
+		allImagesForApp = (app) -> _.map(app.services, imageForService)
+
+		currentImages = _.flatten(_.map(current.local?.apps, allImagesForApp))
```

Comment:
What happens if `current.local` is undefined and how can this happen?

Reply:
Should never happen really, so fixed